brave/brave-ios
# Fix #2210: UserAgent refactor, start using Safari's UA

## Summary of Changes

This pull request fixes issue #2210 

Sec review https://github.com/brave/security/issues/69

This PR is a complete refactor of User Agent logic.
Changes overview:
- mobile UA changed to match Safari. Now both mobile and desktop UA are based on Safari.
- each tab’s webview uses `customUserAgent` for their UA, this makes code logic simpler
- desktop UA is different for iOS 12 & 13, both are based on User Agent from iOS Safari ‚Request desktop website’ mode
- mobile UA also try to follow Safari closely, this required some hardcoded values and future-proofing for upcoming iOS releases
- setting User Agent is now centralized and happens in 2 places: on web view intialization, and on url navigation
- requesting desktop/mobile UA manually now remembers the domains you set and will keep the overridden value until tab is closed.
- tests now include verifying mobile, desktop and iPhone/iPad user agents.


## Submitter Checklist:


## Test Plan:

This requires using the app browsing various webpages to make sure everything works.

I have also prepared some test scenarios for user agent overrides

sample UA website to test https://www.whatismybrowser.com/detect/what-is-my-user-agent
or just type ‚my user agent’ in google search

^let’s call it UAW to make test plan shorter, 

Safari mobile UA:
1. Open safari, open UAW, save result
2. Open Brave, open UAW, save result
3. both UA should be the same

Safari desktop UA:
1. Open safari, open UAW
2. tap on ‚Aa’ on address bar, tap ‚request desktop website’. save result
3. Open Brave, open UAW
4. Tap on 3dots button -> share with -> desktop site, save result
5. UA should be pretty close, but don’t have to be exactly the same. `Version/XXX` and `Mac OS X 10_XX` might slightly differ depending on your iOS version.

Site overrides - iPhone iOS 13 & 12, iPad iOS 12
1. Open Brave, open UAW in tab 1, verify it’s mobile UA(must contain `Mobile/..` in UA)
2. Open UAW in tab 2, verify both UA are the same(tab1 desktop tab2 mobile)
3. On tab 1 tap share and open in desktop mode
4. Verify that UA changed in tab 1, but stays the same on tab 2
5. Refresh both tabs, verify the UA are still the same
6. On both tabs, tap back button, then forward button to go back to the UAW website, verify UA are still the same
7. Refresh both tabs, verify UAs again
8. Open new tab, visit UAW, verify it is showing mobile UA

Site overrides - iPad 13
1. Open Brave, open UAW in tab 1, verify it’s desktop UA(must contain `Macintosh; Intel` in ua)
2. Open UAW in tab 2, verify both UA are the same
3. On tab 1 tap share and open in mobile mode
4. Verify that UA changed in tab 1, but stays the same on tab 2
5. Refresh both tabs, verify the UA are still the same(tab1 mobile, tab2 desktop)
6. On both tabs, tap back button, then forward button to go back to the UAW website, verify UA are still the same
7. Refresh both tabs, verify UAs again
8. Open new tab, visit UAW, verify it is showing desktop UA

Site overrides - iPad 13 with Desktop mode preference disabled
1. Open Brave, go to settings, turn off ‚Always request desktop mode’
2. open UAW in tab 1, verify it’s mobile UA(must contain `Mobile/..` in UA)
3. Open UAW in tab 2, verify both UA are the same(tab1 ~~desktop~~ mobile tab2 mobile)
4. On tab 1 tap share and open in desktop mode
5. Verify that UA changed in tab 1, but stays the same on tab 2
6. Refresh both tabs, verify the UA are still the same
7. On both tabs, tap back button, then forward button to go back to the UAW website, verify UA are still the same
8. Refresh both tabs, verify UAs again
9. Open new tab, visit UAW, verify it is showing mobile UA

UAW 1: first UA verification website
UAW 2: another UA verification website, must be different url than UAW 1

Site overrides - multiple domains
1. Open Brave, go to UAW 1
2. Request mobile/desktop website
3. In the same tab go to UAW 2
4. Verify that user agent from UAW 1 and UAW 2 are different
5. Request desktop/mobile website on UAW 2
6. Verify that now both UAW 1 and UAW 2 have the same UA
7. Open new tab, open UAW 1 and UAW 2, verify that both these UAW on tab 2 are different from tab 1


## Screenshots:



## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
